jsonrpc/types/types.go

```diff
@@ -337,10 +337,27 @@ type Batch struct {
 	SendSequencesTxHash *common.Hash        `json:"sendSequencesTxHash"`
 	VerifyBatchTxHash   *common.Hash        `json:"verifyBatchTxHash"`
 	Transactions        []TransactionOrHash `json:"transactions"`
+	BatchL2Data         ArgBytes            `json:"batchL2Data"`
 }
 
 // NewBatch creates a Batch instance
-func NewBatch(batch *state.Batch, virtualBatch *state.VirtualBatch, verifiedBatch *state.VerifiedBatch, receipts []types.Receipt, fullTx bool, ger *state.GlobalExitRoot) *Batch {
+func NewBatch(
+	batch *state.Batch,
+	virtualBatch *state.VirtualBatch,
+	verifiedBatch *state.VerifiedBatch,
+	receipts []types.Receipt,
+	fullTx bool,
+	ger *state.GlobalExitRoot,
+) (*Batch, error) {
+	batchL2Data := batch.BatchL2Data
+	if batchL2Data == nil {
+		batchL2dataFromTxs, err := state.EncodeTransactions(batch.Transactions)
+		if err != nil {
+			return nil, fmt.Errorf("error encoding txs into raw data: %w", err)
+		} else {
+			batchL2Data = batchL2dataFromTxs
+		}
+	}
```

Comment:
I think this situation is not possible. I mean if the batchL2Data is empty, it's impossible to have decoded transactions. And I'm not sure about this but wouldn't we get an error if the we try to decode an empty array of txs?

Reply:
`batchL2Data` is only stored once the batch is closed. Therefore, while the batch is WIP there are txs but not batchL2Data.

> And I'm not sure about this but wouldn't we get an error if the we try to decode an empty array of txs?
Will try it

Reply:
Even when the batch is wip the sequencer stores the batchL2Data. You can check it in the state.Batch table
![image](https://github.com/0xPolygonHermez/zkevm-node/assets/11256256/e5b38ebb-68ad-4f7e-9937-5d7391538eab)

Reply:
You're right, I had a misconception about this, glad you've clarified this to me. Fixed